### ISSUE TITLE
[ENH] differentiable skcuda

### ISSUE THREAD

**Issue by edouardoyallon** [kymatio/kymatio#128]
Hi,

So Stéphane would be super happy to have a super fast differentiable scattering... who has the time to do this?(not me, I'm sorry about that) A simple unit-test to compare the gradients with torch backend is enough to validate the approach.

Best,
EO

**Comment by edouardoyallon** [comment]
Linked to https://github.com/kymatio/kymatio/issues/261

**Comment by sixin-zh** [comment]
The following cdgmm code supports the backward on GPU, it can be used in backend_skcuda.py

class cdgmmMul(Function):
    @staticmethod
    def forward(ctx, A, B):
        """
        Complex pointwise multiplication between (batched) tensor A and tensor B.

        Parameters
        ----------
        A : tensor
            input tensor with size (B, C, M, N, 2)
        B : tensor
            B is a complex tensor of size (M, N, 2)
        inplace : boolean, optional
            if set to True, all the operations are performed inplace

        Returns
        -------
        C : tensor
            output tensor of size (B, C, M, N, 2) such that:
            C[b, c, m, n, :] = A[b, c, m, n, :] * B[m, n, :]
        """
        A, B = A.contiguous(), B.contiguous()
        if A.size()[-3:] != B.size():
            raise RuntimeError('The filters are not compatible for multiplication!')
        
        if not iscomplex(A) or not iscomplex(B):
            raise TypeError('The input, filter and output should be complex')

        if B.ndimension() != 3:
            raise RuntimeError('The filters must be simply a complex array!')

        if type(A) is not type(B):
            raise RuntimeError('A and B should be same type!')

        if not A.is_cuda:
            raise RuntimeError('Use the torch backend for cpu tensors!')

        ctx.save_for_backward(A,B)
        
        C = A.new(A.size())
        m, n = B.nelement() // 2, A.nelement() // B.nelement()
        lda = m
        ldc = m
        incx = 1
        handle = torch.cuda.current_blas_handle()
        stream = torch.cuda.current_stream()._as_parameter_
        cublas.cublasSetStream(handle, stream)
        cublas.cublasCdgmm(handle, 'l', m, n, A.data_ptr(), lda, B.data_ptr(), incx, C.data_ptr(), ldc)
        return C
    
    @staticmethod
    def backward(ctx, grad_output):
        A, B = ctx.saved_tensors
        conjA = A.clone()
        conjB = B.clone()
        conjA[:,:,:,:,1] = -A[:,:,:,:,1]
        conjB[:,:,1] = -B[:,:,1]
        m, n = conjB.nelement() // 2, conjA.nelement() // conjB.nelement()
        # n is the B*C
        # m is the M*N
        gradA = conjA.new(conjA.size()) # (n,m), col-major
        gradC = grad_output # (n,m), col-major
        # grad_A = grad_C * conj(B)
        lda = m
        ldc = m
        incx = 1
        handle = torch.cuda.current_blas_handle()
        stream = torch.cuda.current_stream()._as_parameter_
        cublas.cublasSetStream(handle, stream)
        cublas.cublasCdgmm(handle, 'l', m, n, gradC.data_ptr(), lda, conjB.data_ptr(), incx, gradA.data_ptr(), ldc)
        
        # grad_B = sum_n grad_C * conj(A)
        # view grad_C and conjA as one vector of size n*m
        gradB_ = gradC.new(gradC.size()) # mul(gradC,conjA) # (B,C,M,N,2)
        lda = m*n
        ldc = m*n
        incx = 1
        cublas.cublasSetStream(handle, stream)
        cublas.cublasCdgmm(handle, 'l', m*n, 1, gradC.data_ptr(), lda, conjA.data_ptr(), incx, gradB_.data_ptr(), ldc)
        gradB = torch.sum(torch.sum(gradB_,0),0) # (m)
        
        return gradA, gradB


cdgmm = cdgmmMul.apply

**Comment by janden** [comment]
Hi @sixin-zh. Thank you for this code! The easiest way to integrate this into the repository would be for you to submit it as a pull request (PR). To do this, you would first create a new branch from the current master, add the new code, commit, and push the branch to your repository. When you go onto the GitHub website afterwards, the branch should have a button "New Pull Request" that you can push. This way, we can go through all the changes together and see if the tests still pass, and so on. If you want, we can discuss this on the Slack channel.

**Comment by janden** [comment]
Sure. It's just the code that's in [this commit](https://github.com/sixin-zh/kymatio/commit/edc3b46b32fb07f9f3698bcb73ec852a16ef2df7), right?

**Comment by sixin-zh** [comment]
Yes, it is.

**Comment by janden** [comment]
Ok. So the new code doesn't support in-place multiplication. Is there some underlying reason for this? I can try to make it work or have it throw an error if you try to do in-place multiplication with the skcuda backend.

**Comment by edouardoyallon** [comment]
@janden there is no specific reason it doesn't support it. I'll review it when I'll get a chance to have some time..

**Comment by janden** [comment]
Ok. Will make it raise an error for now.

**Comment by janden** [comment]
@sixin-zh, I took your commit and added it to a branch on my repository:

https://github.com/janden/kymatio/tree/diff_skcuda_2d

I then made some more changes to integrate the new code into the standard `cdgmm` function:

https://github.com/janden/kymatio/commit/6df832a08369c1233e6e8edf9902dbaed1ad1e39

Using this branch, I then opened a pull request (PR) on the main repository:

https://github.com/kymatio/kymatio/pull/327

We can now use this to discuss the changes, look at how it passes the test suite, make modifications (if needed), and ultimately merge the code into the `master` branch.

**Comment by sixin-zh** [comment]
I have not thought about how to support inplace, you can try to make it work if needed.

**Comment by janden** [comment]
@sixin-zh I've added a basic test for differentiability of `cdgmm` to #327, but it fails for the new skcuda backend. Could you take a look and see what's happening?
